packages/style-engine/class-wp-style-engine.php

```diff
@@ -293,6 +287,159 @@ protected static function is_valid_style_value( $style_value ) {
+		if ( ! $css_selector || ! isset( static::$stores[ $store_key ] ) ) {
+			return false;
+		}
+		$css_declarations = new WP_Style_Engine_CSS_Declarations( $css_declarations );
+		$stored_css_rule  = static::$stores[ $store_key ]->add_rule( $css_selector );
+		$stored_css_rule->add_declarations( $css_declarations );
+		return true;
+	}
+
+	/**
+	 * Taken from gutenberg_enqueue_block_support_styles()
+	 *
+	 * This function takes care of adding inline styles
+	 * in the proper place, depending on the theme in use.
+	 *
+	 * For block themes, it's loaded in the head.
+	 * For classic ones, it's loaded in the body
+	 * because the wp_head action  happens before
+	 * the render_block.
+	 *
+	 * @param callable $callable A user-defined callback function for a WordPress hook.
+	 * @param int      $priority To set the priority for the add_action.
+	 *
+	 * @see gutenberg_enqueue_block_support_styles()
+	 */
+	protected static function render_styles( $callable, $priority = 10 ) {
+		if ( ! $callable ) {
+			return;
+		}
+		$action_hook_name = 'wp_footer';
+		if ( wp_is_block_theme() ) {
+			$action_hook_name = 'wp_head';
+		}
+		add_action( 'wp_enqueue_scripts', $callable );
+		add_action(
+			$action_hook_name,
+			$callable,
+			$priority
+		);
+	}
+
+	/**
+	 * Fetches, processes and compiles stored styles, then renders them to the page.
+	 */
+	public static function process_and_enqueue_stored_styles() {
+		foreach ( static::$stores as $store_key => $store_instance ) {
+			if ( ! $store_instance ) {
+				continue;
+			}
+
+			$css_rules = $store_instance->get_all_rules();
+
+			if ( empty( $css_rules ) ) {
+				continue;
+			}
+
+			$styles_output = '';
+
+			foreach ( $css_rules as $css_rule ) {
+				$styles_output .= $css_rule->get_css();
+			}
+
+			if ( ! empty( $styles_output ) ) {
+				wp_register_style( $store_key, false, array(), true, true );
+				wp_add_inline_style( $store_key, $styles_output );
+				wp_enqueue_style( $store_key );
+			}
+		}
+	}
+
+	/**
+	 * Returns classnames and CSS based on the values in a block attributes.styles object.
+	 * Return values are parsed based on the instructions in BLOCK_STYLE_DEFINITIONS_METADATA.
+	 *
+	 * @param array $block_styles Styles from a block's attributes object.
+	 * @param array $options      array(
+	 *     'selector'                   => (string) When a selector is passed, `generate()` will return a full CSS rule `$selector { ...rules }`, otherwise a concatenated string of properties and values.
+	 *     'convert_vars_to_classnames' => (boolean) Whether to skip converting CSS var:? values to var( --wp--preset--* ) values. Default is `false`.
+	 * );.
+	 *
+	 * @return array array(
+	 *     'css_declarations' => (array) An array of parsed CSS property => CSS value pairs.
+	 *     'classnames'       => (array) A flat array of classnames.
+	 * );
+	 */
+	public function parse_block_supports_styles( $block_styles, $options ) {
```

Comment:
Now that the compile and parse functionality is split, it might be a change to move all the parsing-related methods and utilities to another class or file.